crypto/types/cache.go

```diff
@@ -0,0 +1,72 @@
+package types
+
+import (
+	lru "github.com/hashicorp/golang-lru/v2"
+)
+
+// SignatureCache is a cache of verified signatures
+type SignatureCache struct {
+	data *lru.Cache[string, []byte]
+}
+
+// NewSignatureCache initializes the signature cache.
+// signature verification is one of the most expensive parts in verification
+// by caching it we avoid needing to verify the same signature multiple times
+func NewSignatureCache() (*SignatureCache, error) {
+	// 500 * (32 + 42) = 37.5KB
+	cache, err := lru.New[string, []byte](500)
+	if err != nil {
+		return nil, err
+	}
+
+	return &SignatureCache{data: cache}, nil
+}
+
+// Get returns the cached signature if it exists
+func (c *SignatureCache) Get(key string) ([]byte, bool) {
+	if !c.validate(key) {
+		return nil, false
+	}
+
+	return c.data.Get(key)
+}
+
+// Add adds a signature to the cache
+func (c *SignatureCache) Add(key string, value []byte) {
```

Comment:
My only gripe with this is that I think `Add` should actually take a `cryptotypes.SIgKey` as an argument instead of a `string`. The reason I recommend this is that the `SignatureCache` doesn't really pertain to a signature cache, the API suggests it can be used for any [string,[]byte] mapping. So I'd recommend either changing the argument.

Note, this isn't to block, but just a suggestion.